### ISSUE TITLE
The dimension of `self._state["nn_input"]["nod_fea"]` in the `EGSEnv` class

### ISSUE THREAD

**Issue by zark721** [binqi-sun/egs#3]
Dear author,
I have a question regarding the vector dimension of `self._state["nn_input"]["nod_fea"]` in the `EGSEnv` class (in dag_sched_env.py).

I personally believe that `self._state["nn_input"]["nod_fea"]` represents the observation information, which is directly fed into the actor network for action calculation based on it.

With this understanding, I have the following question:

1. From the code in `EGSEnv`, it seems that the dimension of the node features generated by the environment (`self._state["nn_input"]["nod_fea"]`) is 4. It makes sense because each node has four kinds of information: wcet, eft, lst, and lateral_width. Therefore, I assume that the observation is a two-dimensional matrix with a size n\*4 (where n is the number of nodes). After the process of actor encoder layers, the output should also be n\*4.
However, from the hyperparameters in [arxiv](https://arxiv.org/html/2308.14647v2/#S8)(at TABLE IV:PPO hyper-parameters.), the `Node embedding dimension` is 64. It conflicts with the dimension of `self._state["nn_input"]["nod_fea"]`.
Did I get something wrong? The dimension of observation generated by the environment is n\*64 or n\*4?

Thank you once again for your support.

**Comment by binqi-sun** [comment]
Hi, node features are mapped to a node feature embedding through an embedding layer, and the node embedding dimension is 64. I hope this answers your question.

**Comment by zark721** [comment]
Is the embedding layer you mentioned equivalent to the Encoder Layer here?
![image](https://github.com/binqi-sun/egs/assets/44084756/b4064c09-5781-443f-bc49-f31c928eb3cd)

**Comment by zark721** [comment]
Or does the embedding layer locate before the actor and critic encoder? Just as in the illustration below? If it is, is it a Linear layer or an MLP?
![image](https://github.com/binqi-sun/egs/assets/44084756/87b9aa18-ee70-4dad-b0d8-902d684f949a)

**Comment by binqi-sun** [comment]
Hi, it's an embedding layer before the encoder, as used in many transformer-based architectures. In our case, we use a [Dense layer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Dense) for node feature embeddings and an [Embedding layer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Embedding) for mask embeddings.

**Comment by zark721** [comment]
Thanks again, but I'm not sure about the following details.

1. The Actor and Critic each have their own Dense Layer as the node embedding layer, right? Should the Dense Layer add an activation function?
2. The `mask embedding` you mentioned is for the spatial encoding in Multi-Head Attention, is that correct?
<img src="https://github.com/binqi-sun/egs/assets/44084756/a2fcda68-c401-4e99-807d-0e2f56643084" alt="Image" height="200"/>

3. What is the purpose of `self._state["nn_input"]["pad_msk"]`?
4. In `class EGSEnv`, there are operations like `padding(self._dag.trans_closure, self.max_n_nodes)`. However, the value of `self.max_n_nodes` is the same as the dimension of `self._dag.trans_closure`, so padding here will not have any effect. I would like to know the purpose of the padding operation and setting the variable `self.max_n_nodes`, it seems like they don't play any role over here.

I am a student studying machine learning and I have read transformer and graphormer articles and some code. If you think I lack of other basic knowledge, please directly throw the related link or keywords to me to save your time, thank you very much.
